String abstraction fixes

tautschnig opened the pull request:
Please review commit-by-commit.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
String abstraction fixes

With return-value removal, string abstraction on return values no longer
picked up the right symbol. Furthermore, assignments could end up with
mixed signed/unsigned values in lhs/rhs. Finally, make sure the
left-hand side of an assignment does not have any type casts.

Enable a collection of existing regression tests that weren't being run
to make sure string-abstraction doesn't break again.

tautschnig committed:
clang-format the newly enabled regression tests

These were moved in the prior commit, make sure we follow current
formatting rules.

tautschnig committed:
String abstraction: avoid need for padding

Make the first field a (byte-sized) c_bool instead of (the single-bit)
bool. Otherwise symex complains about the lack of padding, and thus
non-byte-aligned members.

martin-cs commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -16,7 +16,7 @@ jobs:
           DEBIAN_FRONTEND: noninteractive
         run: |
-          sudo apt-get install -yq gcc gdb g++ maven jq flex bison libxml2-utils ccache
+          sudo apt-get install -yq gcc gdb g++ maven jq flex bison libxml2-utils ccache libc6-dev-i386
```
Ummm... why?  This isn't obvious from the commit message.

martin-cs reviewed (approved):
Generally this seems like a good thing.  Fixing broken functionality and resurrecting the tests for it is definitely good.  I realise that this is pretty hypocritical but ... any chance of some documentation?  Page 38 of an old(?) version of the manual here https://www.cprover.org/cbmc/doc/manual.pdf seems to include some description of what string abstraction does.

tautschnig committed:
String abstraction cleanup: use is_parameter annotation

Do not use a class member to communicate what is/isn't a parameter. To
make sure the is_parameter annotation is consistently in place, include
this check in goto_functiont::validate.

tautschnig committed:
String abstraction: add parameters to all function symbols

Do not restrict the updates to the functions in goto_functions, and
create parameters also for unnamed function parameters. While at it,
also cleanup the arguments vs. parameters terminology. Finally, string
abstraction (as it is currently implemented) really only works for a
complete goto model, the application to individual goto programs won't
really work. Thus deprecate the operator() methods in favour of a new
apply() method.

tautschnig commented:
> Generally this seems like a good thing. Fixing broken functionality and resurrecting the tests for it is definitely good. I realise that this is pretty hypocritical but ... any chance of some documentation? Page 38 of an old(?) version of the manual here https://www.cprover.org/cbmc/doc/manual.pdf seems to include some description of what string abstraction does.

@kroening Would you be willing to contribute some documentation, perhaps by taking the above text and adding it to `goto-programs/README.md`?
